# Pandas Assigment 4

## Part 1

In [1]:
import pandas as pd

So during this assignment we will study a dataset that includes house that were sold in year 2014 and 2015 at King County, US.
The main objective is to extract maximum of variables that are affecting the prices in order to reach the ability to fix a missing price for a house knowing all of its specifications (area, grade, view, location etc..)

#### Summary

The study that we will be doing will be based on the following main points:
- List all the variables that we have in our hand 
- Shortlist the variables that might potentially affect the prices
- Take out the average prices by grouping each of theses parameters separately
- Choose the columns that are proven to affect the price based on our results 
- Find patterns and the degree of influence on each of these variables
- Find a price for a house with a missing price value based on all the previous study

#### Importing the dataset

In [3]:
df = pd.read_csv('kc_house_data.csv')

In [437]:
df.sort_values('id', ascending =True).head(30)

id             date     price  bedrooms  bathrooms  sqft_living  \
2497    1000102  20150422T000000  300000.0         6       3.00         2400   
2496    1000102  20140916T000000  280000.0         6       3.00         2400   
6735    1200019  20140508T000000  647500.0         4       1.75         2060   
8411    1200021  20140811T000000  400000.0         3       1.00         1460   
8809    2800031  20150401T000000  235000.0         3       1.00         1430   
3557    3600057  20150319T000000  402500.0         4       2.00         1650   
18521   3600072  20150330T000000  680000.0         4       2.75         2220   
3200    3800008  20150224T000000  178000.0         5       1.50         1990   
21063   5200087  20140709T000000  487000.0         4       2.50         2540   
4333    6200017  20141112T000000  281000.0         3       1.00         1340   
16715   7200080  20141104T000000  239000.0         4       2.00         1980   
16814   7200179  20141016T000000  150000.0         2       1.00          840   
16815   7200179  20150424T000000  175000.0         2       1.00          840   
8508    7400062  20140521T000000  299800.0         2       1.00          790   
4481    7600057  20140805T000000  520000.0         3       2.00         1410   
20483   7600065  20140605T000000  465000.0         3       2.25         1530   
10421   7600125  20141218T000000  630000.0         5       1.00         3020   
20786   7600136  20140718T000000  411000.0         2       2.00         1130   
8677    9000025  20141203T000000  496000.0         2       1.00         1420   
10544  11200070  20140721T000000  570000.0         3       2.50         1530   
16795  11200290  20140609T000000  546000.0         3       2.50         1530   
12062  11200400  20140923T000000  617000.0         3       2.50         1910   
13565  11300120  20140630T000000  635000.0         3       2.50         3350   
11949  11500240  20150428T000000  872750.0         3       2.50         2870   
437    11500890  20150312T000000  843000.0         3       2.50         3130   
1090   11501160  20140617T000000  837700.0         5       2.75         3010   
6881   11501310  20141121T000000  715000.0         3       3.25         3060   
14954  11501330  20140902T000000  795000.0         3       3.50         3190   
184    11510310  20140905T000000  835000.0         4       2.75         3130   
9978   11510700  20140519T000000  755000.0         4       2.50         2660   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
2497       9373     2.0           0     0  ...      7        2400   
2496       9373     2.0           0     0  ...      7        2400   
6735      26036     1.0           0     0  ...      8        1160   
8411      43000     1.0           0     0  ...      7        1460   
8809       7599     1.5           0     0  ...      6        1010   
3557       3504     1.0           0     0  ...      7         760   
18521      5310     1.0           0     0  ...      7        1170   
3200      18200     1.0           0     0  ...      7        1990   
21063      5001     2.0           0     0  ...      9        2540   
4333      21336     1.5           0     0  ...      5        1340   
16715     10585     1.5           0     0  ...      6        1980   
16814     12750     1.0           0     0  ...      6         840   
16815     12750     1.0           0     0  ...      6         840   
8508       5240     1.0           0     0  ...      6         790   
4481       2700     2.0           0     0  ...      7        1410   
20483      1245     2.0           0     0  ...      9        1050   
10421      4800     2.0           0     0  ...      7        3020   
20786      1148     2.0           0     0  ...      9         800   
8677       4635     2.0           0     0  ...      7        1420   
10544      3296     2.0           0     0  ...      8        1530   
16795      3464     2.0           0     0  ...      8        1530   
12062      448

1- First, let's check all the variables available in our hand:

In [5]:
df.columns #List all the columns in our dataFrame

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

2- Next, along with the price column that is the base of our study, we will choose some of these columns which potentially would affect the pricing:

The columns we will be choosing are the following:

Date, bedrooms, sqft_living, sqft_lot, floors, waterfont, view, grade, condition, yr_built, lat, long

In [75]:
list_columns=['bedrooms', 'sqft_living', 'sqft_lot', 'floors', 
              'waterfont', 'view', 'grade', 'condition', 'yr_built','yr_renovated','lat', 'long'] #Save all the shortlisted column in a separate list

3- Now, we will group out the dataframe by each of these columns and check out the average prices:

3.1 First we start checking the bedrooms and living area size:

In [40]:
df[['price','bedrooms','sqft_living']].groupby(['bedrooms']).mean().round(1).reset_index()

bedrooms      price  sqft_living
0          0   410223.1       1707.8
1          1   317658.0        884.6
2          2   401387.7       1239.8
3          3   466276.6       1805.8
4          4   635564.7       2554.6
5          5   786874.1       3047.5
6          6   825853.5       3284.3
7          7   951447.8       3974.2
8          8  1105076.9       3800.0
9          9   893999.8       3775.0
10        10   820000.0       3706.7
11        11   520000.0       3000.0
12        33   640000.0       1620.0

This table represents the average price and average living area size for each bedrooms number.
Looking at the results, as a first impression, we see that the average prices is going up in relation to the number of bedrooms except for certain cases like number 8 and number 9. However for this case the living area size is lower with 9 bedrooms. Which means that the living area size is more indicative in terms of influence on the prices

3.2 We will do the same for the floors column: 

In [42]:
df[['price','floors','sqft_living']].groupby(['floors']).mean().round(1).reset_index()

floors      price  sqft_living
0     1.0   442219.6       1731.8
1     1.5   559044.9       1907.9
2     2.0   649051.5       2571.3
3     2.5  1061020.9       3145.4
4     3.0   582620.1       1788.3
5     3.5   933937.5       2518.0

Seeing the results, the price is not affected by the number of floors as much as the living area size

3.3: Group by condition

In [47]:
df[['price','condition','grade','sqft_living']].groupby(['condition']).agg({'price' : ['mean'], 'grade' : ['median'],'sqft_living':['mean']}).round(1).reset_index()

condition     price  grade sqft_living
                 mean median        mean
0         1  334431.7    6.0      1216.0
1         2  327316.2    7.0      1410.1
2         3  542097.1    8.0      2149.0
3         4  521300.7    7.0      1951.0
4         5  612577.7    7.0      2022.9

We displayed the average prices grouping by condition to check whether the condition has an influence on the price. Despite seeing an ascending pattern, we added up the grade and living area size to check wether it might be due to other variables. However, despite the fact that the living area sizes are changing, the condition still has an effect to a certain extent which we will clearly define in the upcoming sections

3.4: Group by grade:

In [50]:
df[['price','grade']].groupby(['grade']).mean().round(1).reset_index()

grade      price
0       1   142000.0
1       3   205666.7
2       4   214381.0
3       5   248524.0
4       6   301916.6
5       7   402593.3
6       8   542895.5
7       9   773738.2
8      10  1072347.5
9      11  1497792.4
10     12  2192500.0
11     13  3710769.2

The grade variable is definitely one of the most important criterias on house pricings. As shown in the table, the pattern is very clear and average price is linearly increasing in relation to the grade

3.5: Group by year built:

In [53]:
df[['price','yr_built']].groupby(['yr_built']).mean().round(1).reset_index()

yr_built     price
0        1900  581536.6
1        1901  557108.3
2        1902  673192.6
3        1903  480958.2
4        1904  583867.8
..        ...       ...
111      2011  544648.4
112      2012  527437.0
113      2013  678599.6
114      2014  683792.7
115      2015  759970.9

[116 rows x 2 columns]

Clearly the price does not depend on how old the house is. As shown in the table, the average prices do not seem to be affected. However to be sure on this, there are houses that have been renovated and could have an effect on the price. In order to be sure we will have a look on how many houses have been renovated:

In [60]:
df['yr_renovated'].value_counts()

yr_renovated
0       20699
2014       91
2013       37
2003       36
2005       35
        ...  
1951        1
1959        1
1948        1
1954        1
1944        1
Name: count, Length: 70, dtype: int64

In [64]:
perc=(df['yr_renovated'].value_counts()[0]/df.shape[0])*100 #Percentage of non renovated houses
perc

95.77106371165502

Only 5% of houses have been renovated. So the conclusion already made is proven to be correct.

3.6 Group by view and waterfront

In order not to have biased conclucions, we will add up the area size average to make sure that the area does not interfer on the prices.

In [74]:
df[['price', 'waterfront', 'view',
    'sqft_living', 'sqft_lot']].groupby(['waterfront','view']).mean().reset_index()

waterfront  view         price  sqft_living       sqft_lot
0           0     0  4.966235e+05  1997.761660   14179.237416
1           0     1  8.130549e+05  2570.861027   11891.347432
2           0     2  7.839551e+05  2649.277487   22158.072251
3           0     3  9.646851e+05  3039.201629   34123.415479
4           0     4  1.270713e+06  3412.619565   22720.282609
5           1     1  6.350000e+05  1940.000000  167125.000000
6           1     2  1.842188e+06  3369.125000   34060.000000
7           1     3  1.173605e+06  2485.263158   51986.000000
8           1     4  1.728300e+06  3268.133333   20061.251852

In [71]:
df['view'].value_counts()

view
0    19489
2      963
3      510
1      332
4      319
Name: count, dtype: int64

Most houses have 0 views. So we will do our analysis based on 0 or >0 views. 

Clearly there are differences on the prices.
As for houses with 1,2,3,4 views the prices do not seem to be affected.
For houses with same number of views, and with close area sizes, the waterfront variable does make a difference.
So for the following analysis we will be based on the following:
- Whether a house has 0 or >0 views
- If the house has at least 1 view, the price will increase whether there is a waterfront or not.

3.7 Location:Long and Lat:

 We know for a fact that the location does have an influence on the price, so we will keep it in the updated list of columns for further analysis.

4- Based on our analysis, we will pick up all the variables that are proven to be affecting the price. Thus, the list of columns will have to be updated:

In [474]:
list_columns=['id','price','sqft_living', 'sqft_lot','waterfront', 'view', 'grade', 'condition','lat', 'long'] #Updated list of columns

5- Geographical Location:

In order to figure out the prices of the houses with the variables we have, we are required to specify a fixed area on the map since prices could vary largely depending on the location of the house

We will add the distance function given 2 latitudes and longitudes 

In [80]:
from math import sin, cos, sqrt, atan2
def distance(lat1,lon1,lat2,lon2):
    R = 6373.0
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = (sin(dlat/2))**2 + cos(lat1) * cos(lat2) * (sin(dlon/2))**2
    c = 2 * atan2(sqrt(a), sqrt(1-a))
    distance = R * c
    return distance

We will have a on the location where all the houses are located by finding the min and max longitudes and latitudes:

In [81]:
df[['long','lat']].agg(['min','max'])

long      lat
min -122.519  47.1559
max -121.315  47.7776

We will choose a randow sqare area with a ~ 30 km diameter:
- Longitude between -121.850 and -121.845
- Latitude between 47.5290 and 47.5330

In [333]:
distance(47.4880,-122.845,47.4880,-122.850)

29.7759587284091

In [329]:
distance(47.5290,-122.780,47.5330,-122.780)

25.49199999998587

Let us show a subset of all the houses located in this area in order to have a look on the prices: We will insert the results in a separate dataFrame named 'df_area1'

In [446]:
df_area2=df.loc[(df['long'] > -121.850) & (df['long'] <-121.845) & (df['lat'] >47.5290) & (df['lat'] <47.5330),list_columns]

In [447]:
df_area2.sort_values('price', ascending =True)

price  sqft_living  sqft_lot  waterfront  view  grade  condition  \
9255   368000.0         2080      4307           0     0      7          3   
4602   435000.0         2190      6578           0     0      7          3   
11976  440000.0         2230      5800           0     0      7          3   
10760  470000.0         2480      5082           0     0      7          3   
16150  555000.0         2690      4819           0     0      7          3   

           lat     long  
9255   47.5300 -121.848  
4602   47.5305 -121.847  
11976  47.5308 -121.847  
10760  47.5297 -121.846  
16150  47.5302 -121.849

- Now we have a dataframe of 5 houses with the same view, grade, condition and waterfront.
- These houses are located within the same area.
- By sorting the dataFrame by price in the ascending order, it is very clear that the prices are increasing proportionally with the sqft_living area.

In [320]:
# df1=df.sort_values('long', ascending =False)[['long','lat']]

In [337]:
# with pd.option_context('display.max_rows', None, 'display.max_columns', None):
#     display(df1)

By obtaining these results we can now estimate the prices of the houses located within this area by calculating the price per sqft living:

In [448]:
df_area2['price/sqft']=df_area['price']/df_area['sqft_living']

In [449]:
df_area2

price  sqft_living  sqft_lot  waterfront  view  grade  condition  \
4602   435000.0         2190      6578           0     0      7          3   
9255   368000.0         2080      4307           0     0      7          3   
10760  470000.0         2480      5082           0     0      7          3   
11976  440000.0         2230      5800           0     0      7          3   
16150  555000.0         2690      4819           0     0      7          3   

           lat     long  price/sqft  
4602   47.5305 -121.847  198.630137  
9255   47.5300 -121.848  176.923077  
10760  47.5297 -121.846  189.516129  
11976  47.5308 -121.847  197.309417  
16150  47.5302 -121.849  206.319703

In [450]:
df_area2[['price/sqft']].agg(['min','max','mean'])

price/sqft
min   176.923077
max   206.319703
mean  193.739693

So the prices of the houses located within the fixed area with 0 view, 0 waterfront, grade 7 are between 176$ and 206$ which gives us an average of 193$/sqft 

6- Estimation of prices 

In the final step, we will consider that we have a missing price for a certain house and we want to estimate the fair price based on our previous analysis. 
We will take a random price value from the df_area2 dataFrame and we will insert a missing value instead

In [392]:
import numpy as np
missing_percentage = 0.1  # 10% missing values

In [393]:
df_missing = df_area2.copy()

In [394]:
mask = np.random.rand(*df_missing['price'].shape)<missing_percentage
df_missing['price'][mask] = np.nan

C:\Users\lenovo\AppData\Local\Temp\ipykernel_5836\1215590272.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_missing['price'][mask] = np.nan


In [395]:
df_missing

price  sqft_living  sqft_lot  waterfront  view  grade  condition  \
4602        NaN         2190      6578           0     0      7          3   
9255   368000.0         2080      4307           0     0      7          3   
10760  470000.0         2480      5082           0     0      7          3   
11976  440000.0         2230      5800           0     0      7          3   
16150  555000.0         2690      4819           0     0      7          3   

           lat     long  price/sqft  
4602   47.5305 -121.847  198.630137  
9255   47.5300 -121.848  176.923077  
10760  47.5297 -121.846  189.516129  
11976  47.5308 -121.847  197.309417  
16150  47.5302 -121.849  206.319703

So now we have a missing price on the index no. 11976 with a sqft_living=2230.
Following our study, we have to fill in the price as equal to sqft_living * average price/sqft: 2230 * 193

In [372]:
avg_price_sqft=193 # The average price per sqft

In [396]:
df_filled['price'] = df_missing['price'].fillna(df_missing['sqft_living']*avg_price_sqft)

In [397]:
df_filled

price  sqft_living  sqft_lot  waterfront  view  grade  condition  \
4602   422670.0         2190      6578           0     0      7          3   
9255   368000.0         2080      4307           0     0      7          3   
10760  470000.0         2480      5082           0     0      7          3   
11976  440000.0         2230      5800           0     0      7          3   
16150  555000.0         2690      4819           0     0      7          3   

           lat     long  price/sqft  
4602   47.5305 -121.847  198.630137  
9255   47.5300 -121.848  176.923077  
10760  47.5297 -121.846  189.516129  
11976  47.5308 -121.847  197.309417  
16150  47.5302 -121.849  206.319703

As shown in the table, the filled value is 422670 which is an estimation of the price based on the living area.
The initial price before inserting the missing value was 435000 which gives an acceptable estimation with 3% margin of error.

Example 2:

In this example we will choose another area defined by a rectangle between the 2 following coordinates:
- Longitude between -122.405 and -122.400
- Latitude between 47.6370 and 47.6400


In [414]:
distance(47.6400,-122.405,47.6370,-122.405)

19.119000000000725

In [415]:
distance(47.6400,-122.400,47.6400,-122.405)

27.714429368377395

The area chosen has 27 km between longitudes and 19 km between latitudes

In [416]:
df_area3=df.loc[(df['long'] > -122.405) & (df['long'] <-122.400) & (df['lat'] >47.6370) & (df['lat'] <47.6400),list_columns]

In [417]:
df_area3

price  sqft_living  sqft_lot  waterfront  view  grade  condition  \
1962    605000.0         2660      4500           0     0      7          4   
4524    875000.0         2720      4913           0     1      8          4   
4633   1020000.0         3620      4000           0     0     10          3   
7708    592500.0         1340      5350           0     0      8          3   
18072   705000.0         2320      6755           0     0      8          5   
18516   849000.0         2520      4534           0     0      9          5   
19570   619790.0         1040      5097           0     0      7          4   

           lat     long  
1962   47.6391 -122.403  
4524   47.6384 -122.404  
4633   47.6393 -122.403  
7708   47.6384 -122.403  
18072  47.6398 -122.403  
18516  47.6381 -122.401  
19570  47.6385 -122.401

So this area includes a dataFrame of 7 houses:

The remarkable thing about this area is that it has a variation of grades (between 7 and 10) and condition (between 3 and 5) which is different than the case in the previous area. So in this study we will have the chance to have a look on how these variables are affecting the prices.

Since the grade variable seems like the most influencing we will start by sorting our dataFrame by grade:

In [420]:
df_area3.sort_values('grade', ascending =True)[list_columns]

price  sqft_living  sqft_lot  waterfront  view  grade  condition  \
1962    605000.0         2660      4500           0     0      7          4   
19570   619790.0         1040      5097           0     0      7          4   
4524    875000.0         2720      4913           0     1      8          4   
7708    592500.0         1340      5350           0     0      8          3   
18072   705000.0         2320      6755           0     0      8          5   
18516   849000.0         2520      4534           0     0      9          5   
4633   1020000.0         3620      4000           0     0     10          3   

           lat     long  
1962   47.6391 -122.403  
19570  47.6385 -122.401  
4524   47.6384 -122.404  
7708   47.6384 -122.403  
18072  47.6398 -122.403  
18516  47.6381 -122.401  
4633   47.6393 -122.403

Then we will add a price/sqft column to analyze the prices and see how the grade and contion are affecting them

In [424]:
df_area3['price/sqft']=df_area3['price']/df_area3['sqft_living']

In [429]:
df_area3[['price','sqft_living','sqft_lot','view','grade','condition','price/sqft']].sort_values('grade', ascending =True)

price  sqft_living  sqft_lot  view  grade  condition  price/sqft
1962    605000.0         2660      4500     0      7          4  227.443609
19570   619790.0         1040      5097     0      7          4  595.951923
4524    875000.0         2720      4913     1      8          4  321.691176
7708    592500.0         1340      5350     0      8          3  442.164179
18072   705000.0         2320      6755     0      8          5  303.879310
18516   849000.0         2520      4534     0      9          5  336.904762
4633   1020000.0         3620      4000     0     10          3  281.767956

Now we have a table with the average price of each house and sorted by grade
From this table we can draw few conclusion by taking some examples:

1- The first two houses have the same grade and condition, but there is a very high difference in the price rate (227$/sqft and 595$/sqft).
This shows that despite the fact that both houses are within a same small area, there is an area which is much more expensive than the other. For example the house at index 19570 could located in a downtown or at the center of the city. However the house at index 1962 is located in the suburbs


2- If we take the house at index 18072 (grade 8) and the house at index (grade 9) :
They are probably located within the same area where the price rates are the same. However the grade of the house (from grade 8 to grade 9) can increase the prices by 10%. (303$/sqft and 336$/sqft)

Example 3:

Let's have a look and some of the houses sorted by ID:

In [439]:
df_id=df.sort_values('id', ascending =True)

In [440]:
df_id.head(30)

id             date     price  bedrooms  bathrooms  sqft_living  \
2497    1000102  20150422T000000  300000.0         6       3.00         2400   
2496    1000102  20140916T000000  280000.0         6       3.00         2400   
6735    1200019  20140508T000000  647500.0         4       1.75         2060   
8411    1200021  20140811T000000  400000.0         3       1.00         1460   
8809    2800031  20150401T000000  235000.0         3       1.00         1430   
3557    3600057  20150319T000000  402500.0         4       2.00         1650   
18521   3600072  20150330T000000  680000.0         4       2.75         2220   
3200    3800008  20150224T000000  178000.0         5       1.50         1990   
21063   5200087  20140709T000000  487000.0         4       2.50         2540   
4333    6200017  20141112T000000  281000.0         3       1.00         1340   
16715   7200080  20141104T000000  239000.0         4       2.00         1980   
16814   7200179  20141016T000000  150000.0         2       1.00          840   
16815   7200179  20150424T000000  175000.0         2       1.00          840   
8508    7400062  20140521T000000  299800.0         2       1.00          790   
4481    7600057  20140805T000000  520000.0         3       2.00         1410   
20483   7600065  20140605T000000  465000.0         3       2.25         1530   
10421   7600125  20141218T000000  630000.0         5       1.00         3020   
20786   7600136  20140718T000000  411000.0         2       2.00         1130   
8677    9000025  20141203T000000  496000.0         2       1.00         1420   
10544  11200070  20140721T000000  570000.0         3       2.50         1530   
16795  11200290  20140609T000000  546000.0         3       2.50         1530   
12062  11200400  20140923T000000  617000.0         3       2.50         1910   
13565  11300120  20140630T000000  635000.0         3       2.50         3350   
11949  11500240  20150428T000000  872750.0         3       2.50         2870   
437    11500890  20150312T000000  843000.0         3       2.50         3130   
1090   11501160  20140617T000000  837700.0         5       2.75         3010   
6881   11501310  20141121T000000  715000.0         3       3.25         3060   
14954  11501330  20140902T000000  795000.0         3       3.50         3190   
184    11510310  20140905T000000  835000.0         4       2.75         3130   
9978   11510700  20140519T000000  755000.0         4       2.50         2660   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
2497       9373     2.0           0     0  ...      7        2400   
2496       9373     2.0           0     0  ...      7        2400   
6735      26036     1.0           0     0  ...      8        1160   
8411      43000     1.0           0     0  ...      7        1460   
8809       7599     1.5           0     0  ...      6        1010   
3557       3504     1.0           0     0  ...      7         760   
18521      5310     1.0           0     0  ...      7        1170   
3200      18200     1.0           0     0  ...      7        1990   
21063      5001     2.0           0     0  ...      9        2540   
4333      21336     1.5           0     0  ...      5        1340   
16715     10585     1.5           0     0  ...      6        1980   
16814     12750     1.0           0     0  ...      6         840   
16815     12750     1.0           0     0  ...      6         840   
8508       5240     1.0           0     0  ...      6         790   
4481       2700     2.0           0     0  ...      7        1410   
20483      1245     2.0           0     0  ...      9        1050   
10421      4800     2.0           0     0  ...      7        3020   
20786      1148     2.0           0     0  ...      9         800   
8677       4635     2.0           0     0  ...      7        1420   
10544      3296     2.0           0     0  ...      8        1530   
16795      3464     2.0           0     0  ...      8        1530   
12062      448

Looking at this table, we can draw 2 main conclusions:

1- There are rows with same IDs which means that it is the same house sold 2 times, once each year. Based on this we can take some examples and have an idea about the price increase between 2014 and 2015:

Example 1: ID no.1000102:
This house was sold in 2014 at 280000$ then in 2015 at 300000$ which give a 7% increase.

Example 2: ID no.7200179:
This house was sold in 2014 at 150000$ then in 2015 at 175000$ which give a 8.5% increase.

Following the same logic we can take a sample of rows with same house IDs and calculate the average of the price increase between 2014 and 2015

2- Let's have a look and the following table:

In [443]:
df_id[['id','lat','long']].head(30)

id      lat     long
2497    1000102  47.3262 -122.214
2496    1000102  47.3262 -122.214
6735    1200019  47.4444 -122.351
8411    1200021  47.4434 -122.347
8809    2800031  47.4783 -122.265
3557    3600057  47.5803 -122.294
18521   3600072  47.5801 -122.294
3200    3800008  47.4938 -122.262
21063   5200087  47.5423 -122.302
4333    6200017  47.4023 -122.273
16715   7200080  47.4836 -122.214
16814   7200179  47.4840 -122.211
16815   7200179  47.4840 -122.211
8508    7400062  47.5303 -122.288
4481    7600057  47.6029 -122.302
20483   7600065  47.6018 -122.297
10421   7600125  47.6025 -122.313
20786   7600136  47.6023 -122.314
8677    9000025  47.6800 -122.304
10544  11200070  47.6181 -122.138
16795  11200290  47.6179 -122.141
12062  11200400  47.6176 -122.140
13565  11300120  47.7277 -122.207
11949  11500240  47.6944 -122.102
437    11500890  47.6954 -122.103
1090   11501160  47.6960 -122.102
6881   11501310  47.6971 -122.101
14954  11501330  47.6968 -122.102
184    11510310  47.6993 -122.102
9978   11510700  47.6972 -122.104

By looking at this table, we can notice that the houses with close ID numbers are located close to each others.

Example 1: ID 1200019 and 1200021:

We will use the function defined in one of the previous assignments, which calculates the distance based on the IDs

In [444]:
def distance_id(ID1,ID2,df):
    lat1=df.loc[(df['id']==ID1),'lat'].values[0]
    long1=df.loc[(df['id']==ID1),'long'].values[0]
    lat2=df.loc[(df['id']==ID2),'lat'].values[0]
    long2=df.loc[(df['id']==ID2),'long'].values[0]
    return distance(lat1,long1,lat2,long2)

In [445]:
distance_id(1200019,1200021,df_id)

25.02298629824039

The distance is 25 km which is relatively close

Based on this conclusion, we can use an alternative method which is to locate an area by IDs instead of coordinates. This method might give us a better insight about the price rates in every area

### Conclusion

During this assignment, we had a target which was to find the main variables that have a direct effect on the prices in order to estimate a missing price value in the dataFrame.

Based on our study we found a list of columns that when grouping by could show a clear pattern on the prices, so we selected these variables and inserted them in a list named list_columns

Then we took a sample of 2 different areas by precising a square/rectange with a small diameter and extracted the houses within these areas. After then we calculated the average price/sqft and compared these values by looking at the other variables that might affect the prices. 

The first area we chose, had the same variables and the average of prices /sqft seemed stable. So we were able to insert missing values and filled them with the average price found and proved to be more or less accurate 

Finally, we could notice from sorting the houses by ID that there is a pattern between the IDs and the geographical location of the houses. In addition to that, there are houses that were sold twice between 2014 and 2015 and based on that we could get an idea about the price increase between those 2 years.

## Part 2:

The correlation function is used to identify the correlation between all the dataFrame columns. A correlation equals to 1 indicates a 100% intersection which is most probably the same column:

In [475]:
df[list_columns].corr()

id     price  sqft_living  sqft_lot  waterfront      view  \
id           1.000000 -0.016797    -0.012258 -0.132109   -0.002721  0.011592   
price       -0.016797  1.000000     0.702044  0.089655    0.266331  0.397346   
sqft_living -0.012258  0.702044     1.000000  0.172826    0.103818  0.284611   
sqft_lot    -0.132109  0.089655     0.172826  1.000000    0.021604  0.074710   
waterfront  -0.002721  0.266331     0.103818  0.021604    1.000000  0.401857   
view         0.011592  0.397346     0.284611  0.074710    0.401857  1.000000   
grade        0.008130  0.667463     0.762704  0.113621    0.082775  0.251321   
condition   -0.023783  0.036392    -0.058753 -0.008958    0.016653  0.045990   
lat         -0.001891  0.306919     0.052529 -0.085683   -0.014274  0.006157   
long         0.020799  0.021571     0.240223  0.229521   -0.041910 -0.078400   

                grade  condition       lat      long  
id           0.008130  -0.023783 -0.001891  0.020799  
price        0.667463   0.036392  0.306919  0.021571  
sqft_living  0.762704  -0.058753  0.052529  0.240223  
sqft_lot     0.113621  -0.008958 -0.085683  0.229521  
waterfront   0.082775   0.016653 -0.014274 -0.041910  
view         0.251321   0.045990  0.006157 -0.078400  
grade        1.000000  -0.144674  0.114084  0.198372  
condition   -0.144674   1.000000 -0.014941 -0.106500  
lat          0.114084  -0.014941  1.000000 -0.135512  
long         0.198372  -0.106500 -0.135512  1.000000

The variables were the correlation is high are the most variables that affect the prices:
As shown in the table for example, the sqft_living has a correlation of 0.7 with the prices, which indicates that it has a high influce on the price. This is not the same in case of the sqft_lot area.